### ISSUE TITLE
[ENHANCEMENTS] Logrotation / Compression

### ISSUE THREAD

**Issue by Morethanevil** [tobychui/zoraxy#411]
This enhancement shall save storage space. I took a look at the logs, and it seems Zoraxy does not compress them, they get bigger over time. No rotation, just archiving. On my homeserver they are 300MB, but on my main server the log folder it was over 15GB :)

Maybe Zoraxy could compress them with zip or tar. An option in the settings to delete them after X days would be an idea too.

![Screenshot 2024-12-03 232545](https://github.com/user-attachments/assets/90e43fc1-be38-4170-94f9-bcf003dfb941)

**Comment by tobychui** [comment]
@Morethanevil Wow your homelab is really handling some big traffics. 
I am currently planning to reconstruct the whole statistic system with some other faster DB (LevelDB maybe?) so more details can be seen on the statistic analysis & system logger. I could optimize the log file management logic together with this update.

**Comment by Morethanevil** [comment]
That would be cool. An option to delete logs after a month or x days would be cool too.

My main server handles a lot of traffic because it federates with others. Mastodon, Sharkey, Lemmy, Matrix etc... This is normal for me 😄

Most logs by nginx or apache are compressed via tar.gz, maybe Zoraxy could use this too. The newest way in Linux are [binary logs](https://www.howtogeek.com/499623/how-to-use-journalctl-to-read-linux-system-logs/) (journalctl) 

Just a few thoughts

**Comment by LeonardoIz** [comment]
Related to this, it would also be good to add an option that automatically cleans the logs every so often.

**Comment by barto95100** [comment]
Yes the same and possibility button to clear all file or listed file :
![image](https://github.com/user-attachments/assets/a58fff5d-cb1d-4128-a91b-5cfc6caa0e94)

**Comment by flymia** [comment]
@Morethanevil In the meantime, you could setup a custom logrotate rule to handle this. Other reverse proxies/web servers also use logrotate to do that kind of stuff.

Since Zoraxy does not install into default locations and everything is handled in one main directory, writing a logrotate rules and shipping them with Zoraxy is extremely difficult. I would have made a pull request already, but the fact that Zoraxy stores logs in it's own directory and there is (as far as I know) no way to change it, every setup is too custom.

**Comment by tobychui** [comment]
@flymia The Zoraxy logger module do have a parameter for setting the output log location, but I didn't expose it as I didn't thought of this will be necessary. If exposing such interface will make your change / PR easier, please feel free to let me know (with details on what changes can benefit your implementation the most). For easy changes like this, I probably can implement and ship it in the next release :)

**Comment by flymia** [comment]
@tobychui Basically what's missing is an option to set the various paths (not only log). So it would be a fantastic if we could have an option to setup the paths how we want them to be.

This is my zoraxy folder today. I installed it in {{/opt/zoraxy}}:
```
ls -l /opt/zoraxy/
total 130M
drwxr-xr-x. 7 root root  118 Dec 10 13:00 conf/
drwxr-xr-x. 2 root root   28 Dec 10 13:00 log/
-rw-------. 1 root root 132K Dec 29 00:00 sys.db
-rwxr-xr-x. 1 root root   36 Dec 10 13:00 sys.uuid*
drwxr-xr-x. 2 root root   48 Dec 17 10:37 tmp/
drwxr-xr-x. 4 root root   35 Dec 10 13:00 www/
-rwxr-xr-x. 1 root root 130M Nov 24 08:04 zoraxy*
```

It would be great to setup it the way, that I can put the directories into the FHS standard ones, if running natively on the machine (and not in a docker container).

* conf -> /etc/zoraxy
* log -> /var/log/zoraxy
* tmp -> /var/tmp/zoraxy
* www -> /var/www/html (or something like that)

That way I could also make another pull request to setup SELinux policies and build an RPM package that users can use to install it faster and easier natively on their Red Hat based machines. Plus, I can put the systemd service file in the RPM, so that this is also done automatically. Since zoraxy is already running as {{root}} file system permissions are not a big problem.

**Comment by Morethanevil** [comment]
Bringing this up again. For now I am using ZSTD (`sudo apt install zstd`) on my system for compressing the logs. I do this via cron every 30 days.

Adding this to the crontab:
 
`sudo crontab -e`

```bash

0 0 */30 * * zstd -r --rm -T0 /path/to/zoraxy/log >/dev/null 2>&1
```
Operators are: -r for recursive, --rm for removing the logfile after successful compression, -T0 for autodetect CPU cores for multithreading. Replace `*/30` with `*/10` if you want this every 10 days.

Maybe autodeletion would be something as a plugin

Edit: I forgot uncomressing :)

`zstd -d LOGFILE`

**Comment by Morethanevil** [comment]
I wrote a rule for the integrated logrotation of Linux. I put it in [the wiki](https://github.com/tobychui/zoraxy/wiki/FAQ---Frequently-Asked-Questions#logrotation)

**Comment by tobychui** [comment]
New startup flags has been added for v3.2.6 pre-release

```
  -enablelogcompress
        Enable log compression for rotated log files (default true)
  -logrotate int
        Enable log rotation and set the maximum log file size in KB (e.g. 25 for 25KB), set to 0 for disable
```

Since I didn't really have any nodes to test it yet (homelab still work in progress), I will keep this open and see if there are any issue reports come in within a month. If everything works out I guess this can be closed in mid Oct.

**Comment by Morethanevil** [comment]
I did a quick test with those parameters:


`/apps/zoraxy/zoraxy_linux_amd64 -port=:8400 -fastgeoip=true -enablelogcompress=true -logrotate=2500`

And another with this:

`/apps/zoraxy/zoraxy_linux_amd64 -port=:8400 -fastgeoip=true -enablelogcompress=true -logrotate 2500`

`du -sh log` still shows 10MB

What am I doing wrong or does Zoraxy rotate at midnight?

Can we set a specific value for logrotate? Like

`-logrotate 10MB` or `-logrotate 250kB` ?

**Comment by tobychui** [comment]
The current log rotation ticker is set to 1hr per tick (i.e. check and rotate log every 1 hour). 
Alternatively, there is a debug API `/api/log/rotate/debug.trigger`  where you can manually trigger the rotation check if needed.

**Comment by Morethanevil** [comment]
Checked again, now the log folder has 9MB. Is the rotation command `-logrotate=2500` or `-logrotate 2500` for 2500kB?

**Comment by tobychui** [comment]
@Morethanevil I think both will work? This start flag is provided by golang native flag package and the official doc say these syntax in theory should works

```
-flag
--flag   // double dashes are also permitted
-flag=x
-flag x  // non-boolean flags only
```

See more at https://pkg.go.dev/flag#hdr-Command_line_flag_syntax

Yes, I think it should be a easy upgrade to add support for human readable formats like `-logrotate 10MB`.

**Comment by Morethanevil** [comment]
Thanks, this helps a lot with human readable formats 👍🏻

**Comment by Anderhar** [comment]
Stuck to figure out the best approach for compatibility with Crowdsec and other log parsers.

The problem is that Crowdsec [uses](https://github.com/tobychui/zoraxy/discussions/338#discussioncomment-13158507) a regular expression `zr_*.log` to find log file in the target directory. But I suspect this will lead to a conflict when several matching files (`zr_2026-2.log`, `zr_2026-3.log` etc.) accumulate in the same directory, right? (Actually, I tried to simulate this by pressing the “Rotate Log Now” button, but after that I still have only one file `zr_2026-2.log` in the log directory, so can't say for sure.)

How should I configure Zoraxy's log behavior to safely prevent such a conflict, @tobychui ?
